pkg/http/api.go

```diff
@@ -208,249 +208,251 @@ func (a *api) MarkStatusAsOutboundReady() {
+			Version:         apiVersionV1,
+			FastHTTPHandler: a.onDeleteState,
 		},
 		{
-			Methods: []string{nethttp.MethodPost, nethttp.MethodPut},
-			Route:   "state/{storeName}/bulk",
-			Version: apiVersionV1,
-			Handler: a.onBulkGetState,
+			Methods:         []string{nethttp.MethodPost, nethttp.MethodPut},
+			Route:           "state/{storeName}/bulk",
+			Version:         apiVersionV1,
+			FastHTTPHandler: a.onBulkGetState,
 		},
 		{
-			Methods: []string{nethttp.MethodPost, nethttp.MethodPut},
-			Route:   "state/{storeName}/transaction",
-			Version: apiVersionV1,
-			Handler: a.onPostStateTransaction,
+			Methods:         []string{nethttp.MethodPost, nethttp.MethodPut},
+			Route:           "state/{storeName}/transaction",
+			Version:         apiVersionV1,
+			FastHTTPHandler: a.onPostStateTransaction,
 		},
 		{
-			Methods: []string{nethttp.MethodPost, nethttp.MethodPut},
-			Route:   "state/{storeName}/query",
-			Version: apiVersionV1alpha1,
-			Handler: a.onQueryStateHandler(),
+			Methods:         []string{nethttp.MethodPost, nethttp.MethodPut},
+			Route:           "state/{storeName}/query",
+			Version:         apiVersionV1alpha1,
+			FastHTTPHandler: a.onQueryStateHandler(),
 		},
 	}
 }
 
 func (a *api) constructPubSubEndpoints() []Endpoint {
 	return []Endpoint{
 		{
-			Methods: []string{nethttp.MethodPost, nethttp.MethodPut},
-			Route:   "publish/{pubsubname}/{topic:*}",
-			Version: apiVersionV1,
-			Handler: a.onPublish,
+			Methods:         []string{nethttp.MethodPost, nethttp.MethodPut},
+			Route:           "publish/{pubsubname}/*",
+			Version:         apiVersionV1,
+			FastHTTPHandler: a.onPublish,
 		},
 		{
-			Methods: []string{nethttp.MethodPost, nethttp.MethodPut},
-			Route:   "publish/bulk/{pubsubname}/{topic:*}",
-			Version: apiVersionV1alpha1,
-			Handler: a.onBulkPublish,
+			Methods:         []string{nethttp.MethodPost, nethttp.MethodPut},
+			Route:           "publish/bulk/{pubsubname}/*",
+			Version:         apiVersionV1alpha1,
+			FastHTTPHandler: a.onBulkPublish,
 		},
 	}
 }
 
 func (a *api) constructBindingsEndpoints() []Endpoint {
 	return []Endpoint{
 		{
-			Methods: []string{nethttp.MethodPost, nethttp.MethodPut},
-			Route:   "bindings/{name}",
-			Version: apiVersionV1,
-			Handler: a.onOutputBindingMessage,
+			Methods:         []string{nethttp.MethodPost, nethttp.MethodPut},
+			Route:           "bindings/{name}",
+			Version:         apiVersionV1,
+			FastHTTPHandler: a.onOutputBindingMessage,
 		},
 	}
 }
 
 func (a *api) constructDirectMessagingEndpoints() []Endpoint {
 	return []Endpoint{
 		{
-			Methods:           []string{router.MethodWild},
-			Route:             "invoke/{id}/method/{method:*}",
-			Alias:             "{method:*}",
-			Version:           apiVersionV1,
-			KeepParamUnescape: true,
-			Handler:           a.onDirectMessage,
+			// No method is defined here to match any method
+			Methods: []string{},
+			Route:   "invoke/*",
```

Comment:
Before the handler could be invoked in 2 ways:

- `invoke/{id}/method/{method:*}`
- `{method:*}`

Because of that and especially with the introduction of external service invocation, which makes it so `method` can include `/` even in variant 1 (and those can, but don't necessarily have to, be encoded as `%2F`), it made the parsing of the arguments a lot more complex. You can see how even in the current master, the matching is not straightfoward, and we've had bugs in there already.

Because the 1st option is essentially a subset of the 2nd, and because of changes in the router and how it handles %2F etc, the best way to make this work was to remove all path matching decisions from the router and move that to the handler itself.

I did discuss this with @artursouza and @yaron2 on discord, as the possible solution to make everything work, including making URL-encoding of `/` for external service invocation optional.

I'm not saying this is "simple" by all means, as it involves even a state machine (which however has full unit test coverage), but it was either this or a regular expression, which may be simpler or not (depends who you ask :) ) but it's certainly slower.